### PR TITLE
Fix a bug in implementation of rank

### LINKED ISSUE #604: A bug in rank() function in package LinearAlgebra
``` julia
julia> using LinearAlgebra
julia> df=ones(200,2);
julia> rank(df)
1
julia> df=ones(500,2);
julia> rank(df)
2
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, haswell)
Environment:
  JULIAPATH = /home/sunsh/software/julia-1.1.0/bin
```

### LINKED ISSUE #30913: A bug in rank() function in package LinearAlgebra
``` julia
julia> using LinearAlgebra
julia> df=ones(200,2);
julia> rank(df)
1
julia> df=ones(500,2);
julia> rank(df)
2
```

### PR DESCRIPTION
**Pull request by eulerkochy**
Made the implementation same as that of numpy.  This fixes JuliaLang/LinearAlgebra.jl#604

### CODE REVIEW

File: stdlib/LinearAlgebra/test/generic.jl [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 module TestGeneric
 
-using Test, LinearAlgebra, Random
+using Test, LinearAlgebra, Random, Distributions
```
Comment (oxinabox) [inline stdlib/LinearAlgebra/test/generic.jl]: You can't use `Distributions` as it is not a stdlib

File: stdlib/LinearAlgebra/test/generic.jl [hunk 2]
```diff
@@ -179,6 +179,42 @@ end
...
+@test rank(Id,atol=0.99e-8,rtol=0.99e-8) == 4
+@test rank(Id,1.01e-8) == 3
+@test rank(Id,atol=1.01e-8) == 3
+@test rank(Id,rtol=1.01e-8) == 3
+@test rank(Id,atol=1.01e-8,rtol=1.01e-8) == 3
+
+@testset "test_rank" begin
+    nd = Normal()
+    for i = 1:10
+        X = rand(nd, 40, 10)
```
Comment (oxinabox) [inline stdlib/LinearAlgebra/test/generic.jl]:
```suggestion
        X = randn(40, 10)
```

File: stdlib/LinearAlgebra/test/generic.jl [hunk 3]
```diff
@@ -179,6 +179,42 @@ end
...
+@test rank(Id,atol=0.99e-8) == 4
+@test rank(Id,rtol=0.99e-8) == 4
+@test rank(Id,atol=0.99e-8,rtol=0.99e-8) == 4
+@test rank(Id,1.01e-8) == 3
+@test rank(Id,atol=1.01e-8) == 3
+@test rank(Id,rtol=1.01e-8) == 3
+@test rank(Id,atol=1.01e-8,rtol=1.01e-8) == 3
+
+@testset "test_rank" begin
+    nd = Normal()
```
Comment (oxinabox) [inline stdlib/LinearAlgebra/test/generic.jl]:
```suggestion
```

File: stdlib/LinearAlgebra/test/generic.jl (hunk 1 above)
Comment (oxinabox) [reply stdlib/LinearAlgebra/test/generic.jl]:
```suggestion
using Test, LinearAlgebra, Random
```

Comment (eulerkochy) [reply stdlib/LinearAlgebra/test/generic.jl]: Way around for that? I'm using normally random values in a matrix for a test!

Comment (oxinabox) [reply stdlib/LinearAlgebra/test/generic.jl]: `randn` returns normally distributed values

Comment (eulerkochy) [reply stdlib/LinearAlgebra/test/generic.jl]: Cool, just googled about it! Thanks!

**Comment by eulerkochy** [comment]
@oxinabox is it fine now?

**Comment by oxinabox** [comment]
could be.
My comments before were just because you wanted to know why your tests were failing.
I don't have merge rights, and I'd not read the original issue.

**Comment by andreasnoack** [comment]
I don't think using the largest dimension is a good idea. We need to think a little more carefully about a better function of `m` and `n` to use for the default tolerance. As an example of why I don't think the largest dimension is correct, consider a `randn(n)*ones(1, 2)` matrix as `n` grows. Here are the two tolerances and the smallest singular value.
skaermbillede 2019-02-25 kl 11 36 42

Hence, you'd have something like
```julia
julia> A = randn(10000, 2)*Diagonal([1, 1e-12]);

julia> sum(svdvals(A)[1]*eps()*size(A, 1) .< svdvals(A))
1

julia> sum(svdvals(A)[1]*eps()*size(A, 2) .< svdvals(A)) # current rank
2
```

**Comment by eulerkochy** [comment]
Wait, I just tried out the same thing in `numpy`. It also throws the same kind of error. The problem will always persist if the default tolerance value stays close to the entries of the matrix, because of `eps(real(float(one(eltype(A))))`. What do you suggest as a workaround for that? And also how much precision is desired in our calculation, because according to the current implementation, there will always be a trade-off between precision and accuracy?

**Comment by eulerkochy** [comment]
By the way, `MATLAB` also gives incorrect result, and perhaps it's because of the reason I cited above.

**Comment by eulerkochy** [comment]
Only `Wolfram-alpha` is giving an accurate result, so far!

**Comment by oxinabox** [comment]
> Only `Wolfram-alpha` is giving an accurate result, so far!

Yes, I feel like Mathematica is the gold standard for correctness.
We use it in a few other places to generate reference data for tests.

**Comment by andreasnoack** [comment]
> By the way, `MATLAB` also gives incorrect result

I'm not sure it's so easy to decide what is correct and what is incorrect here. It might be that there will always be cases that are not intuitive when it comes to the numerical rank.

> What do you suggest as a workaround for that?

We'll have to understand the get a better understanding of how `m` and `n` influence the result of `rank`. However, it might be very dependent on the properties of the matrix.

> Only `Wolfram-alpha` is giving an accurate result, so far!

What do we know about the algorithm? Wolfram might be computing it in higher precision which would then be a different function from what we are aiming for here.

**Comment by eulerkochy** [comment]
Can we use `BigFloat` to work around this issue?

**Comment by simonbyrne** [comment]
> Can we use `BigFloat` to work around this issue?

No: we don't have have an `svdvals` that works with `BigFloat`s in Base (you need [GenericSVD.jl](https://github.com/JuliaLinearAlgebra/GenericSVD.jl)), and it would be _very_ slow and memory intensive.

**Comment by eulerkochy** [comment]
Have you guys thought of a workaround? I tried working around with some combinations of function, they all seem to fail on some precision or the other :cry:

**Comment by andreasnoack** [comment]
It's just a hard problem. We probably need help from an expert. Maybe @higham would know more about LAPACK's formulation

> modestly growing function of `m` and `n`

mentioned in https://github.com/JuliaLang/julia/issues/30913#issuecomment-467564263

**Comment by oscardssmith** [comment]
Closing as this PR doesn't get us much closer to the answer.

**Comment by rajraomichigan** [comment]
So a random "round off error" matrix with i.i.d $\pm epsilon$ numbers would have largest singular value on the order of $(\sqrt{m} + \sqrt{n}) \epsilon \leq C \sqrt{\max{m,n}} \epsilon  $ with high probability.

This is something @alanedelman  worked on for his PhD thesis

http://math.mit.edu/~edelman/publications/eigenvalues_and_condition_numbers.pdf

Moreover from [this paper](https://arxiv.org/pdf/1003.2990.pdf) and concentration properties of the extreme singular values of such  a random matrix (regardless of whether they are Gaussian or not, broadly speaking) we have that

$ \textrm{Prob}( |\sigma_1(X) - \mathbb{E}[sigma_1](X) | \geq \delta ) \leq C exp(- \delta^2)$

Thus setting threshold to $\max{m,n} \epsilon$ ensures that the probability of exceeding threshold dies down really fast as $m,n$ increase.

That's one justification I can think of.  More generally, the exact threshold for largest singular of the "round off" part  of the spectrum will depend on the original (not rounded off ) matrix structure  -- the bound is the best we can do.

This is how I've thought of the calculation (I might be off in the tail probability calculations). Maybe @alanedelman   can chime in.  

TL; DR the extra $\sqrt{\max{m,n}}$ factor gives it a buffer. Replacing with $\min{m,n}$ seems shaky based on the bounds especially when $m, n = 1$.
